File: 2015/Day06/chatGptSolution.py
def read_instructions(file_name):
    with open(file_name, 'r') as file:
        instructions = [line.strip() for line in file.readlines()]
    return instructions


def apply_instructions(instructions):
    grid = [[0 for _ in range(1000)] for _ in range(1000)]

    for instruction in instructions:
        parts = instruction.split()
        if parts[0] == "toggle":
            action = "toggle"
            start_coords = tuple(map(int, parts[1].split(',')))
            end_coords = tuple(map(int, parts[3].split(',')))
        else:
            action = parts[1]
            start_coords = tuple(map(int, parts[2].split(',')))
            end_coords = tuple(map(int, parts[4].split(',')))

        for i in range(start_coords[0], end_coords[0] + 1):
            for j in range(start_coords[1], end_coords[1] + 1):
                if action == "on":
                    grid[i][j] = 1
                elif action == "off":
                    grid[i][j] = 0
                else:  # toggle
                    grid[i][j] = 1 - grid[i][j]

    return sum(sum(row) for row in grid)


instructions = read_instructions("input.txt")
result = apply_instructions(instructions)
print(result)

File: 2015/Day06/chatGptSolution.hs
import System.IO

data Action = TurnOn | TurnOff | Toggle deriving (Show, Eq)
data Instruction = Instruction Action (Int, Int) (Int, Int) deriving (Show)

parseInstruction :: String -> Instruction
parseInstruction s =
  let ws = words s
      action = case head ws of
                 "turn" -> if ws !! 1 == "on" then TurnOn else TurnOff
                 "toggle" -> Toggle
      coordPairs = map (\x -> read ( "("++x++")" ) :: (Int, Int)) $ filter (\x -> ',' `elem` x) ws
  in Instruction action (head coordPairs) (last coordPairs)

applyInstruction :: Instruction -> [[Bool]] -> [[Bool]]
applyInstruction (Instruction action (x1, y1) (x2, y2)) grid =
  [ [ updateLight action (x, y) light
    | (light, x) <- zip row [0..]]
  | (row, y) <- zip grid [0..]]
  where
    updateLight a (x, y) l
      | x >= x1 && x <= x2 && y >= y1 && y <= y2 = case a of
          TurnOn -> True
          TurnOff -> False
          Toggle -> not l
      | otherwise = l

countLitLights :: [[Bool]] -> Int
countLitLights = sum . map (length . filter id)

main :: IO ()
main = do
  contents <- readFile "input.txt"
  let instructions = map parseInstruction $ lines contents
      initialGrid = replicate 1000 $ replicate 1000 False
      finalGrid = foldl (flip applyInstruction) initialGrid instructions
  print $ countLitLights finalGrid
